### DISCUSSION TITLE
Menu/navigation (like conversations)

### DISCUSSION THREAD

**Discussion by bilogic** [botman/botman#1394] [Q&A]
Has anyone come up with a structured way to implement a menu/navigation system?

e.g. press 1 for aaa, press 2 for bbb, press 0 to return...

**Comment by filippotoso** [comment]
For any driver that supports buttons, I use them for navigation.

For voice bots, I use DTMF and STT, the first one on the ID of the button, the second one on the label of the button.

I wrote my own trait to easily create questions:

```php
<?php

namespace FilippoToso\Botman\Helpers\Traits;

use BotMan\BotMan\Messages\Outgoing\Question;
use BotMan\BotMan\Messages\Outgoing\Actions\Button;
use Illuminate\Support\Arr;

trait Questionable
{
    protected function question($message, $buttons)
    {
        $payload = [];
        foreach ($buttons as $value => $button) {
            if (is_a($button, Button::class)) {
                $payload[] = $button;
            } elseif (is_array($button)) {
                $payload[] = Button::create($button['label'])
                    ->value($value)
                    ->additionalParameters(Arr::except($button, ['label']));
            } else {
                $payload[] = Button::create($button)
                    ->value($value);
            }
        }

        return Question::create($message)->addButtons($payload);
    }
}
```

I pass to it the question (i.e What do you need) and an associative array for the buttons.

**Reply by bilogic** [reply]
nice, let me try it thanks!

**Reply by bilogic** [reply]
It helped, what still seems to be missing is the context, i.e. it remembers which part of the (nested) menu user was in and allows the user to navigate around.

**Reply by filippotoso** [reply]
Use a Conversation and different method for each item in the menu, plus a "back" button to call the previous method and go back in the navigation.

Otherwise, you can use a database (or another data storage) and follow this method: https://github.com/botman/botman/discussions/1387#discussioncomment-7453813
